File: morphological_parser_sak/md.pl
#!/usr/bin/perl
#Author: Haşim Sak @ Department of Computer Engineering - Boğaziçi University.
#Email: hasim.sak@gmail.com
#Version: 2.0 (This version works with the Boun Morphological Parser output)
#Date: May 27, 2009
#Description: This program implements an averaged perceptron based morphological disambiguation system for Turkish text.
#             You can find the most up to date version at http://www.cmpe.boun.edu.tr/~hasim.
#             For more information and license, you can see readme_license.txt and read the following paper.
#             Haşim Sak, Tunga Güngör, and Murat Saraçlar. Morphological disambiguation of Turkish text with perceptron algorithm.
#             In CICLing 2007, volume LNCS 4394, pages 107-118, 2007.
#             If you want to use this program in your research, please cite this paper.
#             Please note that this implementation is a little bit different than the one described in the paper as explained in the readme_license.txt.
#Important: The input files are expected to be UTF-8 encoded. You can use iconv unix command line tool for encoding conversions.

use strict;
use utf8;
use open ":utf8";
use Net::WebSocket::Server;

$| = 1;

binmode STDIN, ":utf8";
binmode STDOUT, ":utf8";

my $num_examples;
my %w;
my %avgw;
my %counts;

# INITIALIZATION
srand(792349);
load_model("model.txt");

Net::WebSocket::Server->new(
    listen => 34215,
    on_connect => sub {
        my ($serv, $conn) = @_;
        $conn->on(
            utf8 => sub {
                my ($conn, $msg) = @_;
                my $disamb_text = disamb($msg);
                $conn->send_utf8($disamb_text);
            },
        );
    },
)->start;


sub disamb
{
	my ($amb_text) = @_;

	my @words;
	my @allparses;

	my $result = "";
	my @lines = split /\n/, $amb_text;
	for my $line (@lines)
	{
		chomp($line);
		if ($line =~ /<DOC>/ || $line =~ /<\/DOC>/ || $line =~ /<TITLE>/ || $line =~ /<\/TITLE>/ || $line =~ /<S>/)
		{
			$result .= "$line\n";
			next;
		}
		if ($line =~ /<\/S>/)
		{
			my ($best_score, @best_parse) = best_parse(1, @allparses);
			for (my $i = 0; $i < @words; ++$i)
			{
				$result .= "$words[$i] $best_parse[$i]";
				my @parses = split(/\s+/, $allparses[$i]);
				foreach my $p (@parses)
				{
					if ($p ne $best_parse[$i])
					{
						$result .= " $p";
					}
				}
				$result .= "\n";
			}
			$result .= "$line\n";
			@words = ();
			@allparses = ();
			next;
		}
		my @tokens = split(/\s+/, $line);
		push(@words, shift(@tokens));
		push(@allparses, "@tokens");
	}
	return $result;
}

sub best_parse
{
	my ($averaged_perceptron, @allparses) = @_;
	push(@allparses, "</s>");
	my %bestpath;
	$bestpath{0} = "-1 0 null";
	my $best_state_num = 0;
	my $best_score;
	my %states;
	$states{"<s> <s>"} = 0;
	my $n = 0;
	foreach my $str (@allparses)
	{
		$best_score = -100000;
		my %next_states;
		my @cands = split(/\s+/, $str);
		#shuffle candidates
		for (my $j = 0; $j < @cands; ++$j)
		{
			my $k = rand(@cands);
			my $temp = $cands[$j];
			$cands[$j] = $cands[$k];
			$cands[$k] = $temp;
		}

		foreach my $cand (@cands)
		{
			foreach my $state (keys %states)
			{
				my $state_num = $states{$state};
				my ($prev, $score, $parse) = split(/\s+/, $bestpath{$state_num});

				my @hist = split(/\s+/, $state);
				my @trigram = ($hist[0], $hist[1], $cand);
				my %feat;
				extract_trigram_feat(\%feat, @trigram);
				my $trigram_score;
				if ($averaged_perceptron)
				{
					$trigram_score = ascore(\%feat);
				}
				else
				{
					$trigram_score = score(\%feat);
				}				
				my $new_score = $score + $trigram_score;

				my $new_state = "$hist[1] $cand";
				if (!defined $next_states{$new_state})
				{
					$next_states{$new_state} = ++$n;
				}
				my $next_state_num = $next_states{$new_state};

				if (defined $bestpath{$next_state_num})
				{
					my ($ignore, $cur_score, $ignore) = split(/\s+/, $bestpath{$next_state_num});					
					if ($new_score > $cur_score)
					{
						$bestpath{$next_state_num} = "$state_num $new_score $cand";
					}
				}
				else
				{
					$bestpath{$next_state_num} = "$state_num $new_score $cand";					
				}
				if ($new_score > $best_score)
				{
					$best_score = $new_score;
					$best_state_num = $next_state_num;
				}
			}
		}
		%states = %next_states;
	}

	my @best_parse;
	my $state_num = $best_state_num;
	while ($state_num != 0)
	{
		my ($prev, $score, $parse) = split(/\s+/, $bestpath{$state_num});
		unshift(@best_parse, $parse);
		$state_num = $prev;
	}
	#pop </s>
	pop(@best_parse);
	return ($best_score, @best_parse);
}

sub extract_features
{
	my ($feat_hash, @parse) = @_;

	unshift(@parse, "<s>");
	unshift(@parse, "<s>");
	push(@parse, "</s>");

	my $i;
	for ($i = 2; $i < @parse; ++$i)
	{
		my @trigram = ($parse[$i-2], $parse[$i-1], $parse[$i]);
		extract_trigram_feat($feat_hash, @trigram);
	}
}

sub extract_trigram_feat
{	
	my ($feat_hash, @trigram) = @_;
	my ($w1, $w2, $w3) = @trigram;

	$trigram[0] =~ s/([\+\-][^\[\]]+\[)/ $1/g;
	my ($r1, @IG1) = split(/\s/, $trigram[0]);
	my $IG1 = join("", @IG1);

	$trigram[1] =~ s/([\+\-][^\[\]]+\[)/ $1/g;
	my ($r2, @IG2) = split(/\s/, $trigram[1]);
	my $IG2 = join("", @IG2);

	$trigram[2] =~ s/([\+\-][^\[\]]+\[)/ $1/g;
	my ($r3, @IG3) = split(/\s/, $trigram[2]);
	my $IG3 = join("", @IG3);

	$feat_hash->{"1:$w1 $w2 $w3"}++;
	$feat_hash->{"2:$w1 $w3"}++;
	$feat_hash->{"3:$w2 $w3"}++;
	$feat_hash->{"4:$w3"}++;
	$feat_hash->{"5:$w2 $IG3"}++;
	$feat_hash->{"6:$w1 $IG3"}++;
	$feat_hash->{"7:$r1 $r2 $r3"}++;
	$feat_hash->{"8:$r1 $r3"}++;
	$feat_hash->{"9:$r2 $r3"}++;
	$feat_hash->{"10:$r3"}++;
	$feat_hash->{"11:$IG1 $IG2 $IG3"}++;
	$feat_hash->{"12:$IG1 $IG3"}++;
	$feat_hash->{"13:$IG2 $IG3"}++;
	$feat_hash->{"14:$IG3"}++;
	foreach my $IG (@IG3)
	{
		$feat_hash->{"15:$IG"}++;
	}
	for (my $j = 0; $j < @IG3; ++$j)
	{
		$feat_hash->{"16:$j $IG3[$j]"}++;
	}
	my $len = @IG3;
	$feat_hash->{"17:$len"}++;
}

sub score
{
	my ($feat_hash) = @_;

	my $score = 0;
	foreach my $feat (keys %{$feat_hash})
	{
		$score += $w{$feat} * $feat_hash->{$feat};
	}
	return $score;
}

sub ascore
{
	my ($feat_hash) = @_;

	my $score = 0;
	foreach my $feat (keys %{$feat_hash})
	{
		$score += $avgw{$feat} * $feat_hash->{$feat};
	}
	return $score;
}


sub load_model
{
	my ($model) = @_;
	open(MODEL, $model) || die("cannot open file: $model\n");
	my $line;
	%avgw = ();
	while ($line = <MODEL>)
	{
		chomp($line);
		$line =~ /([^\s]*)\s+(.*)/;
		my $weight = $1;
		my $feat = $2;
		$avgw{$feat} = $weight;
	}
	close(MODEL);
}

File: morphological_parser/md.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import re
import random
from collections import defaultdict

MODEL_RE = re.compile(r"([^\s]*)\s+(.*)")
FEAT_SEP_RE = re.compile(r"([\+\-][^\[\]]+\[)")
avgw = defaultdict(int)

def disamb(amb_text,addAll):
    """
    amb_test: str
    """
    words = []
    all_parses = []
    result = ""
    lines = [line for line in amb_text.split("\n") if line]
    for line in lines:
        if (("<DOC>" in line) or ("</DOC>" in line) or 
                ("<TITLE>" in line) or ("</TITLE>" in line)):
            result += line + "\n";
            continue
        if "</S>" in line:
            best_score, best_parse_arr = best_parse(all_parses)
            for i, word in enumerate(words):
                result += "{} {}".format(word, best_parse_arr[i])
                if addAll:
                    parses = re.split(r"\s+", all_parses[i])
                    for parse in parses:
                        if parse != best_parse_arr[i]:
                            result += " {}".format(parse)
                result += "\n"
            result += "{}\n".format(line);
            words = []
            all_parses = []
            continue

        word, *parses = re.split(r"\s+", line)
        words.append(word)
        all_parses.append(" ".join(parses))

    return result

def best_parse(all_parses):
    all_parses.append("</s>")
    bestpath = defaultdict(str)
    states = defaultdict(int)
    states["<s> <s>"] = 0
    bestpath[0] = "-1 0 null"
    best_state_num = 0
    n = 0
    for parse in all_parses:
        best_score = -100000
        next_states = dict()
        cands = re.split(r"\s+", parse)
        random.shuffle(cands)
        for cand in cands:
            for state in states.keys():
                state_num = states[state]
                prev, score, parse = re.split(r"\s+", bestpath[state_num])
                score = float(score)
                prev = float(prev)
                hist = re.split(r"\s+", state)
                trigram = [hist[0], hist[1], cand]
                feat = defaultdict(int)
                extract_trigram_feat(feat, trigram)
                trigram_score = ascore(feat)
                new_score = score + trigram_score
                new_state = "{} {}".format(hist[1], cand)
                if new_state not in next_states:
                    n += 1
                    next_states[new_state] = n
                next_state_num = next_states[new_state]
                if next_state_num in bestpath:
                    _, cur_score, _ = re.split(r"\s+", bestpath[next_state_num])
                    cur_score = float(cur_score)
                    if new_score > cur_score:
                        bestpath[next_state_num] = "{} {} {}".format(state_num, new_score, cand)

                else:
                    bestpath[next_state_num] = "{} {} {}".format(state_num ,new_score ,cand)

                if new_score > best_score:
                    best_score = new_score
                    best_state_num = next_state_num
        states = next_states

    best_parse_arr = []
    state_num = best_state_num
    while state_num != 0:
        prev, score, parse = re.split(r"\s+", bestpath[state_num])
        prev = float(prev)
        best_parse_arr.insert(0, parse)
        state_num = prev

    best_parse_arr.pop()
    return [best_score, best_parse_arr]

def extract_trigram_feat(feat_dict, trigram):
    """
    trigram is list
    """
    w1, w2, w3 = trigram
    
    trigram[0] = FEAT_SEP_RE.sub(r" \1", trigram[0])
    r1, *IG1_arr = re.split(r"\s", trigram[0])
    IG1 = "".join(IG1_arr)

    trigram[1] = FEAT_SEP_RE.sub(r" \1", trigram[1])
    r2, *IG2_arr = re.split(r"\s", trigram[1])
    IG2 = "".join(IG2_arr)

    trigram[2] = FEAT_SEP_RE.sub(r" \1", trigram[2])
    r3, *IG3_arr = re.split(r"\s", trigram[2])
    IG3 = "".join(IG3_arr)

    feat_dict["1:{} {} {}".format(w1, w2, w3)] += 1
    feat_dict["2:{} {}".format(w1, w3)] += 1
    feat_dict["3:{} {}".format(w2, w3)] += 1
    feat_dict["4:{}".format(w3)] += 1
    feat_dict["5:{} {}".format(w2, IG3)] += 1
    feat_dict["6:{} {}".format(w1, IG3)] += 1
    feat_dict["7:{} {} {}".format(r1, r2, r3)] += 1
    feat_dict["8:{} {}".format(r1, r3)] += 1
    feat_dict["9:{} {}".format(r2, r3)] += 1
    feat_dict["10:{}".format(r3)] += 1
    feat_dict["11:{} {} {}".format(IG1, IG2, IG3)] += 1
    feat_dict["12:{} {}".format(IG1, IG3)] += 1
    feat_dict["13:{} {}".format(IG2, IG3)] += 1
    feat_dict["14:{}".format(IG3)] += 1

    for ig in IG3_arr:
        feat_dict["15:{}".format(ig)] += 1

    for i, ig in enumerate(IG3_arr):
        feat_dict["16:{} {}".format(i, ig)] += 1

    feat_dict["17:{}".format(len(IG3_arr))] += 1


def ascore(feat_dict):
    global avgw
    score = 0
    for feat in feat_dict.keys():
        score += avgw[feat] * feat_dict[feat]
    return score

def load_model(fname):
    global avgw
    with open(fname, "r") as f:
        lines = [line for line in f.read().split("\n") if line]
    
    for line in lines:
        weight, feat = MODEL_RE.split(line)[1:3]
        avgw[feat] = float(weight)


def init(fname):
    load_model(fname)
